pkg/scheduler/framework/v1alpha1/interface.go

```diff
@@ -0,0 +1,94 @@
+limitations under the License.
+*/
+
+// This file defines the scheduling framework plugin interfaces.
+
+package v1alpha1
+
+import (
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	internalcache "k8s.io/kubernetes/pkg/scheduler/internal/cache"
+)
+
+// Plugin is the parent type for all the scheduling framework plugins.
+type Plugin interface {
+	Name() string
+}
+
+// ReservePlugin is an interface for Reserve plugins. These plugins are called
+// at the reservation point, AKA "assume". These are meant to updated the state
+// of the plugin. They do not return any value (other than error).
+type ReservePlugin interface {
+	Plugin
+	// Reserve is called by the scheduling framework when the scheduler cache is
+	// updated.
+	Reserve(pc *PluginContext, p *v1.Pod, nodeName string) error
+}
+
+// PrebindPlugin is an interface that must be implemented by "prebind" plugins.
+// These plugins are called before a pod being scheduled
+type PrebindPlugin interface {
+	Plugin
+	// Prebind is called before binding a pod. All prebind plugins must return
+	// failure = false, err = nil or the pod will be rejected and won't be sent
+	// for binding. "failure" indicates that an unexpected failure has happened.
+	// Plugins that run successfully and want to reject a pod should return
+	// false, non-nil error.
+	Prebind(pc *PluginContext, p *v1.Pod, nodeName string) (failure bool, err error)
+}
+
+// SchedulerCache is an interface that provides read-only access to the scheduler
+// cache for plugins.
+// NOTE: Scheduler cache contains live data. The cache is snapshotted at the
+// beginning of a scheduling cycle and then the contents of that snapshot is
+// passed to plugins when plugins are called by the framework. If plugins use this
+// interface to read the cache directly, they may receive some information which
+// is different from what is passed to them by the framework.
+type SchedulerCache interface {
+	// IsAssumedPod returns true if the pod is assumed and not expired.
+	IsAssumedPod(*v1.Pod) (bool, error)
+	// List lists all cached pods (including assumed ones).
+	ListPods(labels.Selector) ([]*v1.Pod, error)
+	// GetPod returns the pod from the cache with the same UID as the specified pod.
+	GetPod(*v1.Pod) (*v1.Pod, error)
+}
+
+// Framework manages the set of plugins used by the scheduling framework.
+// Configured plugins are called at specified points in a scheduling cycle.
+type Framework interface {
+	// NodeInfoSnapshot return the latest NodeInfo snapshot. The snapshot
+	// is taken at the beginning of a scheduling cycle and remains unchanged until a pod
+	// finishes "Reserve". There is no guarantee that the information remains unchanged
+	// after "Reserve".
+	NodeInfoSnapshot() *internalcache.NodeInfoSnapshot
+	// SchedulerCache returns an instance of the scheduler with limited functionality.
+	SchedulerCache() SchedulerCache
+	// ReservePlugins returns a slice of "reserve" plugins.
+	ReservePlugins() []ReservePlugin
```

Comment:
Is there a way to pass additional info when registering framework like metrics related information or some arbitrary information?

Reply:
Each plugin can have its own arbitrary config. The framework configuration will be different though. The configuration of the framework itself, should be a part of the core scheduler config.